pkg/controller/installation/products/threescale/reconciler.go

```diff
@@ -0,0 +1,592 @@
+	}
+
+	return v1alpha1.PhaseInProgress, nil
+}
+
+func (r *Reconciler) reconcileOperator(ctx context.Context, serverClient pkgclient.Client) (v1alpha1.StatusPhase, error) {
+	err := r.mpm.CreateSubscription(
+		ctx,
+		serverClient,
+		r.installation,
+		marketplace.GetOperatorSources().Integreatly,
+		r.namespace,
+		packageName,
+		"integreatly",
+		[]string{r.namespace},
+		coreosv1alpha1.ApprovalAutomatic)
+	if err != nil && !k8serr.IsAlreadyExists(err) {
+		return v1alpha1.PhaseFailed, err
+	}
+
+	ip, _, err := r.mpm.GetSubscriptionInstallPlan(ctx, serverClient, packageName, r.namespace)
+	if err != nil && !k8serr.IsNotFound(err) {
+		return v1alpha1.PhaseFailed, err
+	}
+
+	if ip != nil && ip.Status.Phase == coreosv1alpha1.InstallPlanPhaseComplete {
+		return v1alpha1.PhaseCompleted, nil
+	}
+
+	return v1alpha1.PhaseInProgress, nil
+}
+
+func (r *Reconciler) reconcileComponents(ctx context.Context, serverClient pkgclient.Client) (v1alpha1.StatusPhase, error) {
+	bucket := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: s3BucketSecretName,
+		},
+	}
+
+	err := serverClient.Get(ctx, pkgclient.ObjectKey{Name: bucket.Name, Namespace: r.ConfigManager.GetOperatorNamespace()}, bucket)
+	if err != nil {
+		return v1alpha1.PhaseFailed, err
+	}
+
+	s3SecretName := s3CredentialsSecretName
+	tsS3 := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      s3SecretName,
+			Namespace: r.namespace,
+		},
+	}
+	err = serverClient.Get(ctx, pkgclient.ObjectKey{Name: tsS3.Name, Namespace: tsS3.Namespace}, tsS3)
+	if err != nil && k8serr.IsNotFound(err) {
+		// We are copying the s3 details for now but this is not ideal as the secrets can get out of sync.
+		// We need to revise how this secret is set
+		s3 := &v1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: s3SecretName,
+			},
+		}
+		err = serverClient.Get(ctx, pkgclient.ObjectKey{Name: s3.Name, Namespace: r.ConfigManager.GetOperatorNamespace()}, s3)
+		if err != nil {
+			return v1alpha1.PhaseFailed, err
+		}
+
+		tsS3.Data = s3.Data
+		err = serverClient.Create(ctx, tsS3)
+		if err != nil {
+			return v1alpha1.PhaseFailed, err
+		}
+	}
+
+	resourceRequirements := false
+	apim := &threescalev1.APIManager{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      apiManagerName,
+			Namespace: r.namespace,
+		},
+		Spec: threescalev1.APIManagerSpec{
+			APIManagerCommonSpec: threescalev1.APIManagerCommonSpec{
+				WildcardDomain:              r.installation.Spec.RoutingSubdomain,
+				ResourceRequirementsEnabled: &resourceRequirements,
+			},
+			System: &threescalev1.SystemSpec{
+				FileStorageSpec: &threescalev1.SystemFileStorageSpec{
+					S3: &threescalev1.SystemS3Spec{
+						AWSBucket: string(bucket.Data["AWS_BUCKET"]),
+						AWSRegion: string(bucket.Data["AWS_REGION"]),
+						AWSCredentials: v1.LocalObjectReference{
+							Name: s3SecretName,
+						},
+					},
+				},
+			},
+		},
+	}
+	err = serverClient.Get(ctx, pkgclient.ObjectKey{Name: apim.Name, Namespace: r.namespace}, apim)
```

Comment:
Should we check this before loading in the secret?

Reply:
I want to just reconcile on the objects themselves. if secret is not there, reconcile it. if ApiManager is not there reconcile it. etc.

Reply:
if I'm reading the code right, it goes:
1. load in the secret data
2. create the apimanager object in memory
3. see if apimanager exists in api
4. create apimanager if it doesn't exist

I'm suggesting we change the order to:
1. see if apimanager exists, and if not
2.1. load in the secret data
2.2. create the apimanager object in memory
2.3. create apimanager if it doesn't exist

Reply:
I think we should avoid reconciling objects based on whether another object is there.
We should check if the object exists then reconcile that object.